pkg/template/servicebroker/servicebroker.go

```diff
@@ -25,85 +25,112 @@ import (
 // Broker represents the template service broker.  It implements
 // openservicebroker/api.Broker.
 type Broker struct {
+	// privilegedKubeClientConfig is a privileged clientconfig used for communicating back to the
+	// API server.  This is needed because the template service broker is a fake server component that actually
+	// needs both the apiserver and controllers to be running
+	privilegedKubeClientConfig restclient.Config
+	initialize                 sync.Once
+
 	kc                 kclientset.Interface
 	templateclient     internalversiontemplate.TemplateInterface
 	extkc              kclientsetexternal.Interface
 	extrouteclient     extrouteclientset.RouteV1Interface
 	lister             templatelister.TemplateLister
+	hasSynced          func() bool
 	templateNamespaces map[string]struct{}
 	ready              chan struct{}
 }
 
 var _ api.Broker = &Broker{}
 
-// NewBroker returns a new instance of the template service broker.  While
+// DeprecatedNewBrokerInsideAPIServer returns a new instance of the template service broker.  While
 // built into origin, its initialisation is asynchronous.  This is because it is
 // part of the API server, but requires the API server to be up to get its
 // service account token.
-func NewBroker(privrestconfig restclient.Config, privkc kclientset.Interface, infraNamespace string, informer templateinformer.TemplateInformer, namespaces []string) *Broker {
+func DeprecatedNewBrokerInsideAPIServer(privilegedKubeClientConfig restclient.Config, informer templateinformer.TemplateInformer, namespaces []string) *Broker {
 	templateNamespaces := map[string]struct{}{}
 	for _, namespace := range namespaces {
 		templateNamespaces[namespace] = struct{}{}
 	}
 
 	b := &Broker{
+		privilegedKubeClientConfig: privilegedKubeClientConfig,
 		lister:             informer.Lister(),
+		hasSynced:          informer.Informer().HasSynced,
 		templateNamespaces: templateNamespaces,
 		ready:              make(chan struct{}),
 	}
 
-	go func() {
-		// the broker is initialised asynchronously because fetching the service
-		// account token requires the main API server to be running.
+	return b
+}
+
+// MakeReady actually makes the broker functional
+func (b *Broker) MakeReady() error {
+	select {
+	case <-b.ready:
```

Comment:
I'm pretty convinced this isn't correct, but I suppose the response is: it's already merged and it's an intermediate step.
As the code stands, if MakeReady() blocks, all incoming API requests are blocked and never return a response.  The timing code never gets to run; the ready channel is superfluous.